cirq/ops/pauli_string.py

```diff
@@ -1008,63 +1054,292 @@ def _from_json_dict_(  # type: ignore
         return cls(pauli=pauli, qubit=qubit)
 
 
-class _MutablePauliString:
+class MutablePauliString:
+
+    def __init__(self,
+                 *contents: 'cirq.PAULI_STRING_LIKE',
+                 coefficient: Union[int, float, complex] = 1,
+                 pauli_int_dict: Optional[Dict['cirq.Qid', int]] = None):
```

Comment:
I haven't checked the performance, but during multiplication it replaces quite a lot of if-else-if-else-if with `a ^ b`.

I actually do intend to replace the `Qid` type with an `Any` type at a later point, but avoided doing that in this PR because it will also affect `PauliString` pretty significantly.

This parameter isn't so much for users' convenience as it is an internal shortcut to avoid all conversions and double-checks. That's why its type is so specific. I could prefix it with `_` to make it especially clear that it's more of an implementation detail, but for a class driven by performance like this one I think these choices are really part of the API. The dictionary is exposed as a field after all.

Reply:
WDYT about using the shortcut internally while requiring gate objects in the API? I realize this incurs the cost of `isinstance`, but perhaps that's acceptable? Alternatively, we could remove that cost by making [`Pauli._index`](https://github.com/quantumlib/Cirq/blob/master/cirq/ops/pauli_gates.py#L45) public, i.e. `MutablePauliString.__init__` would accept gates, but then it would convert to ints internally by calling `gate.pauli_index` or somesuch (failure on non-Pauli gates being a documented and expected behavior).

Reply:
A large part of the reason that this class exists is because of performance differential. We've had issues in the past w.r.t. equating and hashing custom objects being slow, so I wanted to avoid requiring them.

Reply:
SGTM